### ISSUE TITLE
What is the need for a separate binary type for EH_Frame?

### ISSUE THREAD

**Issue by vishwacs111** [mclinker/mclinker#178]
Hi,
I'm trying to support C++ ABI for a new architecture. I don't understand why we need a separate binary LDFileFormat type for EH_Frame.
I am using DwarfCFI at compiler (llvm). I'm getting .eh_frame section and .eh_frame_hdr sections from compiler. The compiler also generates .rela.eh_frame section. The linker is already using these relocation records to perform relocation on .eh_frame and also allocates addresses. I plan to use the generated eh_frame in libcxxabi during exception handling. But while emitting the section I observe that while reading the EH_Frame from input binary, the fragment list is not getting populated. This is leading to a linker crash.
My question is why does eh_frame need special attention? Why can't it be dealt lke any other constant table present in .data or .strtab using the LDFileFormat Kind Regular.
Thanks in advance

**Comment by DianaChen** [comment]
eh_frame sections, unlike other regular sections, require some special way to read, merge and emit them. In mclinker, we'd like to reduce the size of eh_frame, which is to remove the unused and duplicated  FDE and CIE entries of eh_frame. This work cannot be done by simply analyzing the relocation or section information, but requires to parse all the contents in eh_frame. And also after size reduction, eh_frame is formed to be some special representation in the linker, so a customized way to emit it is required. (The related functions are EhFrameReader::read, EhFrame::merge and ELFObjectWriter::emitEhFrame).  And that's why we make a special LDFileFormat Kind for it, those functions of eh_frame are different from regular sections. I'm not sure I understand what is the problem you encountered. If the eh_frame sections were read in and emitted out by mclinker, then everything  related to section kind should be correctly handled. It would be helpful if you can provide more information about how you use MCLinker. And please let us know if you see any bug. Thank you.

**Comment by vishwacs111** [comment]
Hi Diana,
Sorry for the delay and thanks a lot for your answer.
You are right I need the CIEs and FDEs to be resolved at link time and I understand the need for a separate binary type for EH_frame.
I tried to dig a little more on why mclinker was not able to read the CIEs and FDEs generated by my compiler. I found the following issue.
There are 2 versions of CIE; 1 and 3. In CIE, there is a field to hold the return address register. This field is encoded as an unsigned byte in version1 and as an ULEB128 in version 3. The register number I need to send is bigger than 256. So I had to go with version 3. I believe MCLinker doesn't consider this.
At the location lib/LD/EhFrameReader.cpp:199, to skip the return address register you just increment the char* pointer handler once. I applied the following changes
-  if (cie_end - handler < 1) {
+  if (!skip_LEB128(&handler, cie_end)) {
     return false;
   }
-  ++handler;
After this change the mclinker successfully created the eh_frame and eh_frame_hdr sections.
The issue I'm facing now is even after this change when linker re-constructs the eh_frame, it is updating the version somewhere as 1 again. So I'm getting half the CIEs with version 3 and the other half as version 1 in the linked binary. Can you point me where I can modify this?

**Comment by DianaChen** [comment]
Yes you are correct that we should consider the CIE version when reading CIE return address register. The latest commit (ac5cc111377abe763b563500331fe8246665c3a6) should have fixed it. Thank you very much! For the problem you encountered that there are part of CIEs with version 1, I'm thinking that might because those CIEs are generated by linker for PLT. In MCLinker we generate only version 1 CIEs (The corresponding function is X86_32GNULDBackend::createCIERegionForPLT in X86LDBackend.cpp). Is there any PLT section in your program? Can you help to check if those version 1 CIEs are generated for PLT? Also I'm wondering that will it cause any problem if an eh_frame section contains CIEs with different versions? Thank you.
